### PR TITLE
[WIP] Java time parser

### PR DESCRIPTION
**Pull request by nahi**
It's under review. Do not merge!

### CODE REVIEW

File: embulk-core/src/main/java/org/embulk/spi/time/TimestampParser.java [hunk 1]
```diff
@@ -33,17 +40,20 @@ public TimestampParser(String format, ParserTask task)
...
         this.defaultTimeZone = defaultTimeZone;
     }
 
     public Timestamp parse(String text) throws TimestampParseException
     {
-        long localUsec = helper.strptimeUsec(text);
-        String zone = helper.getZone();
+        Temporal tmp = parser.date_strptime(compiledPattern, text);
+        if (tmp == null) {
+            throw new TimestampParseException();
```
Comment (nahi) [inline embulk-core/src/main/java/org/embulk/spi/time/TimestampParser.java]: tmp cannot be null so we can remove null check, right?

File: embulk-jruby-extension/src/main/java/org/jruby/util/ParsedValues.java [hunk 2]
```diff
@@ -0,0 +1,406 @@
...
+        ZONE_OFFSET_MAP.put("us eastern",             -18000);
+        ZONE_OFFSET_MAP.put("us mountain",            -25200);
+        ZONE_OFFSET_MAP.put("vladivostok",             36000);
+        ZONE_OFFSET_MAP.put("w. australia",            28800);
+        ZONE_OFFSET_MAP.put("w. central africa",        3600);
+        ZONE_OFFSET_MAP.put("w. europe",                3600);
+        ZONE_OFFSET_MAP.put("west asia",               18000);
+        ZONE_OFFSET_MAP.put("west pacific",            36000);
+        ZONE_OFFSET_MAP.put("yakutsk",                 32400);
+    }
```
Comment (nahi) [inline embulk-jruby-extension/src/main/java/org/jruby/util/ParsedValues.java]: You created this definition based on Ruby's Date library, right? Please add source reference as a comment.

File: embulk-jruby-extension/src/main/java/org/jruby/util/ParsedValues.java [hunk 3]
```diff
@@ -0,0 +1,406 @@
...
+                if (len >= 4)
+                    min = parseInt(z.substring(2, 4));
+                if (len >= 6)
+                    sec = parseInt(z.substring(4, 6));
+            }
+        }
+
+        offset = hour * 3600 + min * 60 + sec;
+        return sign ? offset : -offset;
+    }
```
Comment (nahi) [inline embulk-jruby-extension/src/main/java/org/jruby/util/ParsedValues.java]: You ported date_zone_to_diff in CRuby's ext/date, right? Please provide source reference as a comment.

Comment (nahi) [inline embulk-jruby-extension/src/main/java/org/jruby/util/ParsedValues.java]: With the list what you didn't port. ex. DST, Rational.

File: embulk-jruby-extension/src/main/java/org/jruby/util/ParsedValues.java [hunk 4]
```diff
@@ -0,0 +1,406 @@
...
+
+    private static int parseInt(String text)
+    {
+        int v = 0;
+        try {
+            for (int i = 0; i < text.length(); i++) {
+                char c = text.charAt(i); // IndexOutOfBounds
```
Comment (nahi) [inline embulk-jruby-extension/src/main/java/org/jruby/util/ParsedValues.java]: How IndexOOBException occurs here?

File: embulk-jruby-extension/src/main/java/org/jruby/util/RubyDateParser.java [hunk 5]
```diff
@@ -0,0 +1,608 @@
...
+    private static boolean isValidRange(long v, int lower, int upper)
+    {
+        return lower <= v && v <= upper;
+    }
+
+    private static boolean isSpace(char c)
+    {
+        return c == ' ' || c == '\t' || c == '\n' ||
+                c == '\u000b' || c == '\f' || c == '\r';
+    }
```
Comment (nahi) [inline embulk-jruby-extension/src/main/java/org/jruby/util/RubyDateParser.java]: Hmm. I realized that Character.isWhitespace() has different list for detection. It's OK, CRuby might need a fix.

File: embulk-jruby-extension/src/main/java/org/jruby/util/RubyDateParser.java [hunk 6]
```diff
@@ -0,0 +1,608 @@
...
+        numPatterns.add(Format.FORMAT_YEAR_SHORT); // 'y'
+    }
+
+    private static boolean matchAtNumPatterns(Token token) // NUM_PATTERN_P
+    {
+        if (token == null) {
+            return false;
+        }
+
+        // TODO
```
Comment (nahi) [inline embulk-jruby-extension/src/main/java/org/jruby/util/RubyDateParser.java]: Please provide comments for the TODO.

File: embulk-jruby-extension/src/main/java/org/jruby/util/RubyDateParser.java [hunk 7]
```diff
@@ -0,0 +1,608 @@
...
+    }
+
+    private static RuntimeException newInvalidDataException() // Token token, int v
+    {
+        return new RuntimeException("Invalid Data"); // TODO InvalidDataException
+    }
+
+    private final ThreadContext context;
+    // Use RubyDateFormatter temporarily because it has useful lexer, token and format types
+    private final RubyDateFormatter dateFormat;
```
Comment (nahi) [inline embulk-jruby-extension/src/main/java/org/jruby/util/RubyDateParser.java]: Ah, so we need ThreadContext in RubyDateParser. Let's consider the strategy to merge it to JRuby later.

File: embulk-jruby-extension/src/main/java/org/jruby/util/RubyDateParser.java [hunk 8]
```diff
@@ -0,0 +1,608 @@
...
+    {
+        long v = 0;
+        try {
+            for (int i = 0; i < len; i++) {
+                char c = text.charAt(pos); // IndexOutOfBounds
+                if (!isDigit(c)) {
+                    if (i > 0) {
+                        break;
+                    } else {
+                        throw newInvalidDataException();
```
Comment (nahi) [inline embulk-jruby-extension/src/main/java/org/jruby/util/RubyDateParser.java]: Please provide a message for investigation.

File: embulk-core/src/main/java/org/embulk/spi/time/TimestampParser.java (hunk 1 above)
Comment (muga) [inline embulk-core/src/main/java/org/embulk/spi/time/TimestampParser.java]: ok, i removed it. https://github.com/embulk/embulk/commit/cb49a8a

File: embulk-jruby-extension/src/main/java/org/jruby/util/ParsedValues.java (hunk 2 above)
Comment (muga) [inline embulk-jruby-extension/src/main/java/org/jruby/util/ParsedValues.java]: commented. thanks. https://github.com/embulk/embulk/commit/cb49a8a

File: embulk-jruby-extension/src/main/java/org/jruby/util/ParsedValues.java (hunk 3 above)
Comment (muga) [inline embulk-jruby-extension/src/main/java/org/jruby/util/ParsedValues.java]: yes, ported date_strptime.c https://github.com/embulk/embulk/commit/cb49a8a

File: embulk-jruby-extension/src/main/java/org/jruby/util/ParsedValues.java (hunk 4 above)
Comment (muga) [inline embulk-jruby-extension/src/main/java/org/jruby/util/ParsedValues.java]: thanks. i fixed it. https://github.com/embulk/embulk/commit/cb49a8a

File: embulk-jruby-extension/src/main/java/org/jruby/util/RubyDateParser.java (hunk 5 above)
Comment (muga) [inline embulk-jruby-extension/src/main/java/org/jruby/util/RubyDateParser.java]: commented the link of the whitespace declaration: https://github.com/embulk/embulk/commit/cb49a8a

File: embulk-jruby-extension/src/main/java/org/jruby/util/RubyDateParser.java (hunk 6 above)
Comment (muga) [inline embulk-jruby-extension/src/main/java/org/jruby/util/RubyDateParser.java]: removed it. https://github.com/embulk/embulk/commit/cb49a8a

File: embulk-jruby-extension/src/main/java/org/jruby/util/RubyDateParser.java (hunk 7 above)
Comment (muga) [inline embulk-jruby-extension/src/main/java/org/jruby/util/RubyDateParser.java]: Yes, current RubyDateParser has JRuby's RubyDateFormatter.java to use compiled tokens (patterns). When this implementation is merged into JRuby, it should have Lexer and Token definition.

File: embulk-jruby-extension/src/main/java/org/jruby/util/RubyDateParser.java (hunk 8 above)
Comment (muga) [inline embulk-jruby-extension/src/main/java/org/jruby/util/RubyDateParser.java]: This is bug. fixed it. https://github.com/embulk/embulk/commit/cb49a8a

File: embulk-jruby-extension/src/main/java/org/jruby/util/RubyDateParser.java [hunk 9]
```diff
@@ -0,0 +1,605 @@
...
+    }
+
+    static int toInt(char c)
+    {
+        return c - '0';
+    }
+
+    private static Set<Format> numPatterns; // CDdeFGgHIjkLlMmNQRrSsTUuVvWwXxYy
+    static {
+        numPatterns = new HashSet<>();
```
Comment (frsyuki) [inline embulk-jruby-extension/src/main/java/org/jruby/util/RubyDateParser.java]: Use EnumSet

File: embulk-jruby-extension/src/main/java/org/jruby/util/RubyDateParser.java [hunk 10]
```diff
@@ -0,0 +1,605 @@
...
+    {
+        return compilePattern(context.runtime.newString(format), true);
+    }
+
+    public List<Token> compilePattern(RubyString format, boolean dateLibrary)
+    {
+        return dateFormat.compilePattern(format, dateLibrary);
+    }
+
+    private static Token getToken(List<Token> compiledPattern, int index)
```
Comment (frsyuki) [inline embulk-jruby-extension/src/main/java/org/jruby/util/RubyDateParser.java]: Is this method necessary?

File: embulk-jruby-extension/src/main/java/org/jruby/util/RubyDateParser.java [hunk 11]
```diff
@@ -0,0 +1,605 @@
...
+    {
+        return compiledPattern.get(index);
+    }
+
+    private static Token nextToken(List<Token> compiledPattern, int index)
```
Comment (frsyuki) [inline embulk-jruby-extension/src/main/java/org/jruby/util/RubyDateParser.java]: This method is used only with matchAtNumPatterns. Why don't you combine them to one method?

File: embulk-jruby-extension/src/main/java/org/jruby/util/RubyDateParser.java [hunk 12]
```diff
@@ -0,0 +1,605 @@
...
+                    long d;
+                    if (text.charAt(pos) == ' ') { // brank or not
+                        pos += 1; // brank
+                        d = readDigits(1);
+                    } else {
+                        d = readDigits(2);
+                    }
+
+                    if (!isValidRange(d, 1, 31)) {
+                        values.fail();
```
Comment (frsyuki) [inline embulk-jruby-extension/src/main/java/org/jruby/util/RubyDateParser.java]: values need to include some messages why the parsing failed. In CRuby, error message is:

```
irb(main):002:0> Time.strptime('40', '%e')
ArgumentError: invalid strptime format - `%e'
	from /Users/frsyuki/.rbenv/versions/2.2.0/lib/ruby/2.2.0/time.rb:427:in `strptime'
	from (irb):2
	from /Users/frsyuki/.rbenv/versions/2.2.0/bin/irb:11:in `<main>'
```

File: embulk-jruby-extension/src/main/java/org/jruby/util/RubyDateParser.java [hunk 13]
```diff
@@ -0,0 +1,605 @@
...
+            return null;
+        } else {
+            return compiledPattern.get(index + 1);
+        }
+    }
+
+    // TODO RubyTime parse(RubyString format, RubyString text);
+    // TODO RubyTime parse(List<Token> compiledPattern, RubyString text);
+
+    public Temporal date_strptime(List<Token> compiledPattern, String text)
```
Comment (frsyuki) [inline embulk-jruby-extension/src/main/java/org/jruby/util/RubyDateParser.java]: Why do you use snake case method name?

File: embulk-jruby-extension/src/main/java/org/jruby/util/RubyDateParser.java [hunk 14]
```diff
@@ -0,0 +1,605 @@
...
+        this(Ruby.newInstance().getCurrentContext());
+    }
+
+    public RubyDateParser(ThreadContext context)
+    {
+        this.context = context;
+        this.dateFormat = new RubyDateFormatter(context);
+    }
+
+    public List<Token> compilePattern(String format)
```
Comment (frsyuki) [inline embulk-jruby-extension/src/main/java/org/jruby/util/RubyDateParser.java]: Is this method necessary...?

File: embulk-jruby-extension/src/main/java/org/jruby/util/RubyDateParser.java [hunk 15]
```diff
@@ -0,0 +1,605 @@
...
+    private final ThreadContext context;
+    // Use RubyDateFormatter temporarily because it has useful lexer, token and format types
+    private final RubyDateFormatter dateFormat;
+
+    private List<Token> compiledPattern;
+    private int pos;
+    private String text;
+    private ParsedValues values;
+
+    public RubyDateParser()
```
Comment (frsyuki) [inline embulk-jruby-extension/src/main/java/org/jruby/util/RubyDateParser.java]: why did you add this constructor?

File: embulk-jruby-extension/src/main/java/org/jruby/util/Temporal.java [hunk 16]
```diff
@@ -0,0 +1,90 @@
+package org.jruby.util;
+
+import org.joda.time.DateTime;
+import org.joda.time.DateTimeZone;
+
+public class Temporal // TODO better naming
+{
+    // @see https://github.com/jruby/jruby/blob/master/core/src/main/java/org/jruby/RubyTime.java#L1366
```
Comment (frsyuki) [inline embulk-jruby-extension/src/main/java/org/jruby/util/Temporal.java]: This link comment to the master branch with line number doesn't make sense because the line changes when someone changes the code.

File: embulk-jruby-extension/src/main/java/org/jruby/util/ParsedValues.java [hunk 17]
```diff
@@ -0,0 +1,406 @@
+package org.jruby.util;
+
+import java.util.HashMap;
+import java.util.Map;
+
+import static org.jruby.util.RubyDateParser.isDigit;
+import static org.jruby.util.RubyDateParser.isSign;
+import static org.jruby.util.RubyDateParser.toInt;
+
+public class ParsedValues
```
Comment (frsyuki) [inline embulk-jruby-extension/src/main/java/org/jruby/util/ParsedValues.java]: I think putting ParsedValues to org.jruby.util package is not a good idea because the name is generic. Why don't you move this to RubyDateParser.ParsedValues?

File: embulk-jruby-extension/src/main/java/org/jruby/util/ParsedValues.java [hunk 18]
```diff
@@ -0,0 +1,406 @@
...
+
+        boolean sign = false;
+        sign = z.charAt(0) == '+';
+        z = z.substring(1, z.length());
+
+        int hour = 0, min = 0, sec = 0;
+        if (z.contains(":")) {
+            String[] splited = z.split(":");
+            if (splited.length == 2) {
+                hour = parseInt(splited[0]);
```
Comment (frsyuki) [inline embulk-jruby-extension/src/main/java/org/jruby/util/ParsedValues.java]: Why don't you use `Integer.parseInt`?

File: embulk-jruby-extension/test/mri/date/test_date_strptime_simple.rb [hunk 19]
```diff
@@ -0,0 +1,319 @@
+#
+# how to use
+# $ jruby -J-Xbootclasspath/a:embulk-jruby-extension/build/libs/embulk-jruby-extension-0.6.7.jar embulk-jruby-extension/test/mri/date/test_date_strptime_simple.rb
+#
+require 'test/unit'
+require 'date'
+
+require 'java'
+
+class TestDateStrptime < Test::Unit::TestCase
```
Comment (frsyuki) [inline embulk-jruby-extension/test/mri/date/test_date_strptime_simple.rb]: Is this mostly copy of MRI's test/date/test_date_strptime.rb?

File: embulk-jruby-extension/test/mri/date/test_date_strptime_simple.rb [hunk 20]
```diff
@@ -0,0 +1,319 @@
...
+
+  def test__strptime
+    parser = org.jruby.util.RubyDateParser.new
+
+    STRFTIME_2001_02_03.each do |f, s|
+      if (f == '%I' and s[0] == '12') or
+	 (f == '%l' and s[0] == '12') # hour w/o merid
+	s[1]["hour"] = 12
+      end
+      assert_equal(s[1], parser.date_strptime_internal(f, s[0]).toMap.to_hash, [f, s].inspect)
```
Comment (frsyuki) [inline embulk-jruby-extension/test/mri/date/test_date_strptime_simple.rb]: Reusing `parser` instance sounds invalid because it has internal state.

File: embulk-jruby-extension/test/mri/date/test_date_strptime_simple.rb [hunk 21]
```diff
@@ -0,0 +1,319 @@
...
+      case f[-1,1]
+      when 'c', 'C', 'x', 'X', 'y', 'Y'
+	f2 = f.sub(/\A%/, '%E')
+	assert_equal(s[1], parser.date_strptime_internal(f2, s[0]).toMap.to_hash, [f2, s].inspect)
```
Comment (frsyuki) [inline embulk-jruby-extension/test/mri/date/test_date_strptime_simple.rb]: too many toMap.to_hash call...

File: embulk-jruby-extension/test/mri/date/test_date_strptime_simple.rb [hunk 22]
```diff
@@ -0,0 +1,319 @@
...
+# $ jruby -J-Xbootclasspath/a:embulk-jruby-extension/build/libs/embulk-jruby-extension-0.6.7.jar embulk-jruby-extension/test/mri/date/test_date_strptime_simple.rb
+#
+require 'test/unit'
+require 'date'
+
+require 'java'
+
+class TestDateStrptime < Test::Unit::TestCase
+
+  STRFTIME_2001_02_03 = {
```
Comment (frsyuki) [inline embulk-jruby-extension/test/mri/date/test_date_strptime_simple.rb]: Did you change symbolized map to string map like `{"wday"=>6}` to `{:wday=>6}`? I think it's not a good idea because it makes it very difficult to merge upstream changes. We don't want to create a fork.

File: embulk-jruby-extension/src/main/java/org/jruby/util/ParsedValues.java [hunk 23]
```diff
@@ -0,0 +1,406 @@
...
+    int hour = Integer.MIN_VALUE;
+    int year = Integer.MIN_VALUE;
+    int sec = Integer.MIN_VALUE;
+    int wnum0 = Integer.MIN_VALUE;
+    int wnum1 = Integer.MIN_VALUE;
+
+    String zone = null;
+
+    int sec_fraction = Integer.MIN_VALUE; // Rational
+    int sec_fraction_size = Integer.MIN_VALUE;
```
Comment (frsyuki) [inline embulk-jruby-extension/src/main/java/org/jruby/util/ParsedValues.java]: Do you need sec_fraction + sec_fraction_size? Why don't you have nanoseconds only?
`Rational(1, 10) == Rational(10, 100)` is true.

File: embulk-jruby-extension/src/main/java/org/jruby/util/Temporal.java [hunk 24]
```diff
@@ -0,0 +1,90 @@
...
+            if (values.has(values.sec)) {
+                dt = dt.plusSeconds(values.sec);
+            }
+
+            if (values.has(values.sec_fraction)) {
+                nsec = values.sec_fraction * (int)Math.pow(10, 9 - values.sec_fraction_size);
+                dt = dt.plusMillis((int)nsec / 1000000);
+            }
+
+            dt = dt.withZoneRetainFields(dtz);
```
Comment (frsyuki) [inline embulk-jruby-extension/src/main/java/org/jruby/util/Temporal.java]: Does here convert timezone from UTC to UTC? Is this necessary?

File: embulk-jruby-extension/src/main/java/org/jruby/util/Temporal.java [hunk 25]
```diff
@@ -0,0 +1,90 @@
...
+import org.joda.time.DateTime;
+import org.joda.time.DateTimeZone;
+
+public class Temporal // TODO better naming
+{
+    // @see https://github.com/jruby/jruby/blob/master/core/src/main/java/org/jruby/RubyTime.java#L1366
+    public static Temporal newTemporal(ParsedValues values)
+    {
+        long sec;
+        long nsec = 0;
```
Comment (frsyuki) [inline embulk-jruby-extension/src/main/java/org/jruby/util/Temporal.java]: int?

File: embulk-jruby-extension/src/main/java/org/jruby/util/RubyDateParser.java [hunk 26]
```diff
@@ -0,0 +1,605 @@
...
+        }
+        return false;
+    }
+
+    private final ThreadContext context;
+    // Use RubyDateFormatter temporarily because it has useful lexer, token and format types
+    private final RubyDateFormatter dateFormat;
+
+    private List<Token> compiledPattern;
+    private int pos;
```
Comment (frsyuki) [inline embulk-jruby-extension/src/main/java/org/jruby/util/RubyDateParser.java]: This class is not thread safe but public methods don't make us feeling multithread unsafe. Same API design problem with SimpledateFormat. This should be improved.

File: embulk-jruby-extension/src/main/java/org/jruby/util/ParsedValues.java [hunk 27]
```diff
@@ -0,0 +1,406 @@
...
+            return offset;
+        }
+
+        if (z.startsWith("gmt") || z.startsWith("utc")) {
+            z = z.substring(3, z.length()); // remove "gmt" or "utc"
+        }
+
+        if (!isSign(z.charAt(0))) {
+            // if z doesn't have "+" or "-", invalid
+            return offset;
```
Comment (frsyuki) [inline embulk-jruby-extension/src/main/java/org/jruby/util/ParsedValues.java]: Is returning offset here correct? It says it's a invalid offset.

File: embulk-jruby-extension/src/main/java/org/jruby/util/ParsedValues.java (hunk 18 above)
Comment (frsyuki) [inline embulk-jruby-extension/src/main/java/org/jruby/util/ParsedValues.java]: oh, I was confused. This returns Integer.MIN_VALUE. This code is too much C style...

File: embulk-jruby-extension/src/main/java/org/jruby/util/RubyDateParser.java [hunk 28]
```diff
@@ -0,0 +1,605 @@
...
+                case FORMAT_WEEKYEAR: // %G - The week-based year
+                {
+                    long y;
+                    if (matchAtNumPatterns(nextToken(compiledPattern, i))) {
+                        y = readDigits(4);
+                    } else {
+                        y = readDigitsMax();
+                    }
+                    values.cwyear = (int)y;
+                    break;
```
Comment (frsyuki) [inline embulk-jruby-extension/src/main/java/org/jruby/util/RubyDateParser.java]: Is this cast valid? If `readDigitsMax` returns numbers larger than Integer.MAX_VALUE, this cast silently returns broken number.

File: embulk-jruby-extension/test/mri/date/test_date_strptime_simple.rb [hunk 29]
```diff
@@ -0,0 +1,319 @@
...
+      when 'c', 'C', 'x', 'X', 'y', 'Y'
+	f2 = f.sub(/\A%/, '%E')
+	assert_equal(s[1], parser.date_strptime_internal(f2, s[0]).toMap.to_hash, [f2, s].inspect)
+      else
+	f2 = f.sub(/\A%/, '%E')
+	assert_equal(nil, parser.date_strptime_internal(f2, s[0]), [f2, s].inspect)
+	assert_equal({}, parser.date_strptime_internal(f2, f2).toMap, [f2, s].inspect)
+      end
+      case f[-1,1]
+      when 'd', 'e', 'H', 'k', 'I', 'l', 'm', 'M', 'S', 'u', 'U', 'V', 'w', 'W', 'y'
```
Comment (frsyuki) [inline embulk-jruby-extension/test/mri/date/test_date_strptime_simple.rb]: Does this include `'k'`? I copied latest code from MRI and it doesn't include `'k'` here.

File: embulk-jruby-extension/src/main/java/org/jruby/util/ParsedValues.java (hunk 27 above)
Comment (frsyuki) [inline embulk-jruby-extension/src/main/java/org/jruby/util/ParsedValues.java]: Oh, I was confused. This line always returns Integer.MIN_VALUE. Too much C style...

File: embulk-jruby-extension/src/main/java/org/jruby/util/RubyDateParser.java (hunk 7 above)
Comment (frsyuki) [inline embulk-jruby-extension/src/main/java/org/jruby/util/RubyDateParser.java]: Yes. And please try to make RubyDateParser#parseInternal and RubyDateFormatter#format independent from Ruby runtime (org.jruby.Ruby). If they don't need Ruby runtime, we can simplify both embulk-core and plugin code because org.embulk.spi.time.TimestampParser/TimestampFormatter don't need ScriptingContainer.

**Comment by frsyuki** [comment]
@muga Please merge #181.

**Comment by frsyuki** [comment]
For future your work, please keep in mind that semantically unnecessary private fields (such as fail, compiledPattern, etc) are always harmful.

File: embulk-jruby-extension/src/main/java/org/jruby/util/Temporal.java [hunk 30]
```diff
@@ -0,0 +1,90 @@
...
+
+            int year;
+            if (values.has(values.year)) {
+                year = values.year;
+            } else {
+                year = 1970;
+            }
+
+            // set up with min values and then add to allow rolling over
+            DateTime dt = new DateTime(year, 1, 1, 0, 0, 0, 0, DateTimeZone.UTC);
```
Comment (frsyuki) [inline embulk-jruby-extension/src/main/java/org/jruby/util/Temporal.java]: This setting `1970-01-01 00:00:00` is correct behavior? Is this same behavior with CRuby?

File: embulk-jruby-extension/src/main/java/org/jruby/util/Temporal.java [hunk 31]
```diff
@@ -0,0 +1,90 @@
...
+            if (values.has(values.min)) {
+                dt = dt.plusMinutes(values.min);
+            }
+            if (values.has(values.sec)) {
+                dt = dt.plusSeconds(values.sec);
+            }
+
+            if (values.has(values.sec_fraction)) {
+                nsec = values.sec_fraction * (int)Math.pow(10, 9 - values.sec_fraction_size);
+                dt = dt.plusMillis((int)nsec / 1000000);
```
Comment (frsyuki) [inline embulk-jruby-extension/src/main/java/org/jruby/util/Temporal.java]: Is this line necessary?
